src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+        return ForkProject(
+            this.Solution.WithProjectCompilationOutputInfo(projectId, info),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectCompilationOutputInfo"/>
+    public SolutionCompilationState WithProjectDefaultNamespace(
+        ProjectId projectId, string? defaultNamespace)
+    {
+        return ForkProject(
+            this.Solution.WithProjectDefaultNamespace(projectId, defaultNamespace),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectChecksumAlgorithm"/>
+    public SolutionCompilationState WithProjectChecksumAlgorithm(
+        ProjectId projectId, SourceHashAlgorithm checksumAlgorithm)
+    {
+        var tuple = this.Solution.WithProjectChecksumAlgorithm(projectId, checksumAlgorithm);
+        return ForkProject(
+            tuple,
+            new CompilationAndGeneratorDriverTranslationAction.ReplaceAllSyntaxTreesAction(tuple.newProject, isParseOptionChange: false),
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectName"/>
+    public SolutionCompilationState WithProjectName(
+        ProjectId projectId, string name)
+    {
+        return ForkProject(
+            this.Solution.WithProjectName(projectId, name),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectFilePath"/>
+    public SolutionCompilationState WithProjectFilePath(
+        ProjectId projectId, string? filePath)
+    {
+        return ForkProject(
+            this.Solution.WithProjectFilePath(projectId, filePath),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectCompilationOptions"/>
+    public SolutionCompilationState WithProjectCompilationOptions(
+        ProjectId projectId, CompilationOptions options)
+    {
+        var tuple = this.Solution.WithProjectCompilationOptions(projectId, options);
+        return ForkProject(
+            tuple,
+            new CompilationAndGeneratorDriverTranslationAction.ProjectCompilationOptionsAction(tuple.newProject, isAnalyzerConfigChange: false),
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectParseOptions"/>
+    public SolutionCompilationState WithProjectParseOptions(
+        ProjectId projectId, ParseOptions options)
+    {
+        var tuple = this.Solution.WithProjectParseOptions(projectId, options);
+
+        if (this.PartialSemanticsEnabled)
+        {
+            // don't fork tracker with queued action since access via partial semantics can become inconsistent (throw).
+            // Since changing options is rare event, it is okay to start compilation building from scratch.
+            return ForkProject(
+                tuple,
+                translate: null,
+                forkTracker: false);
```

Comment:
the one place where we don't fork the tracker.

Reply:
This threw me a bit, but not beacuse of anything you did -- it seems old code was quite strange here. forkTracker = true says "keep the tracker around and fork it with the optional translation", and forkTracker = false means "just throw it away". Sure, this one place needs a throwaway (but as mentioned, this seems realy suspicious), but oddly there's a lot of places that are creating a new SolutionCompilationState for things that don't impact a compilation at all like the WithProjectName or WithProjectPath, etc and we're just going to be doing extra work finalizing those compilations.

As a tracking for a follow up, it seems either we should make this have a third state: "KeepCompilation", "ForkToUpdateCompilation", or "ThrowAwayCompilation" (the last one being unique here). Or if we can get rid of this per the comment applying a few lines earlier, then that last one goes away and we can have a better use here.

But you're 100% keeping the behavior the same -- the old code is just confusing me now. :smile: